datafusion/physical-expr/src/intervals/cp_solver.rs

```diff
@@ -0,0 +1,1088 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Constraint propagator/solver for custom PhysicalExpr graphs.
+
+use std::collections::HashSet;
+use std::fmt::{Display, Formatter};
+use std::ops::Index;
+use std::sync::Arc;
+
+use arrow_schema::DataType;
+use datafusion_common::{Result, ScalarValue};
+use datafusion_expr::Operator;
+use petgraph::graph::NodeIndex;
+use petgraph::stable_graph::{DefaultIx, StableGraph};
+use petgraph::visit::{Bfs, Dfs, DfsPostOrder, EdgeRef};
+use petgraph::Outgoing;
+
+use crate::expressions::{BinaryExpr, CastExpr, Literal};
+use crate::intervals::interval_aritmetic::{apply_operator, Interval};
+use crate::utils::{build_dag, ExprTreeNode};
+use crate::PhysicalExpr;
+
+// Interval arithmetic provides a way to perform mathematical operations on
+// intervals, which represent a range of possible values rather than a single
+// point value. This allows for the propagation of ranges through mathematical
+// operations, and can be used to compute bounds for a complicated expression.
+// The key idea is that by breaking down a complicated expression into simpler
+// terms, and then combining the bounds for those simpler terms, one can
+// obtain bounds for the overall expression.
+//
+// For example, consider a mathematical expression such as x^2 + y = 4. Since
+// it would be a binary tree in [PhysicalExpr] notation, this type of an
+// hierarchical computation is well-suited for a graph based implementation.
+// In such an implementation, an equation system f(x) = 0 is represented by a
+// directed acyclic expression graph (DAEG).
+//
+// In order to use interval arithmetic to compute bounds for this expression,
+// one would first determine intervals that represent the possible values of x
+// and y. Let's say that the interval for x is [1, 2] and the interval for y
+// is [-3, 1]. In the chart below, you can see how the computation takes place.
+//
+// This way of using interval arithmetic to compute bounds for a complex
+// expression by combining the bounds for the constituent terms within the
+// original expression allows us to reason about the range of possible values
+// of the expression. This information later can be used in range pruning of
+// the provably unnecessary parts of `RecordBatch`es.
+//
+// References
+// 1 - Kabak, Mehmet Ozan. Analog Circuit Start-Up Behavior Analysis: An Interval
+// Arithmetic Based Approach, Chapter 4. Stanford University, 2015.
+// 2 - Moore, Ramon E. Interval analysis. Vol. 4. Englewood Cliffs: Prentice-Hall, 1966.
+// 3 - F. Messine, "Deterministic global optimization using interval constraint
+// propagation techniques," RAIRO-Operations Research, vol. 38, no. 04,
+// pp. 277{293, 2004.
+//
+// ``` text
```

Comment:
I really like this writeup / description

I think it would be great to consolidate the other range analysis in the codebase around this one